### DISCUSSION TITLE
[FirebaseAuth] Expose startActivityForSignInWithProvider

### DISCUSSION THREAD

**Discussion by sageata** [firebase/flutterfire#5560] [Feature request]
**Is your feature request related to a problem? Please describe.**

I am trying to use Sign in With Apple in a Flutter project, however I find it very difficult to implement this on Android because the `startActivityForSignInWithProvider` is not exposed.

**Describe the solution you'd like**
Expose `startActivityForSignInWithProvider` on the FirebaseAuth.instance

**Additional context**

**Comment by Salakar** [comment]
We plan to support the custom OAuth flow (webview / browser signin) in a future release which uses this method underneath on Android for example. No ETA yet though sorry, will probably be after our rework of Functions, Storage, Crashlytics & Messaging.

**Reply by rahulraj64** [reply]
Waiting for this. Any ETA ?

**Reply by lukeirvin** [reply]
To verify, can we use Firebase Apple Auth with Flutter right now?

**Reply by michaelsof47** [reply]
is it resolve? because i find any dependency for my flutter project. but, it can't found a dependency work well.

**Comment by ZetiMente** [comment]
Why is this such a low priority? Is this because its Apple? Here we are in Feb 2021 and you still can't use Firebase Auth for Android using Sign in with Apple. Granted the number of users that would use this is low, it's not zero. I didn't even plan to use Sign in with Apple, but Apple rejects your App if you use other 3rd party logins. So I have no choice but to use it. But now I do all this work for iOS, and can't even enjoy all the fruits of that labor to use the Flutter cross platform power to have Android because of this issue.

**Comment by KevinKozj** [comment]
+1 on @ZetiMente

**Comment by gabrc52** [comment]
Apple sign in brings some benefits like not having to give the app your email address, but still receive emails. I'd like to try it out on my Android device, mainly for testing, since I can't easily test on Apple devices

**Comment by vbuberen** [comment]
Sign in with Apple is also very useful for seamless experience for users with devices on both platforms. 

This issue should get a higher priority for sure.

**Comment by DennisZachotzky** [comment]
Any news on this?

**Comment by gabrc52** [comment]
Definitely, if you use an iPad and an Android phone, you could use Apple Sign in on FlutterFire apps in both.

**Comment by loicgeek** [comment]
+1

**Comment by rafipanoyan** [comment]
Waiting for updates on this too

**Comment by LittleChit-cmd** [comment]
Any news on this?

**Comment by Klabauterman** [comment]
Will you merge it, if I make a PR? I need this as well.

**Comment by dancout** [comment]
This is not directly addressing the issue of exposing `startActivityForSignInWithProvider` for Android, but I was able to find an alternative way to sign in using Apple on an Android device.

* I followed the Configure Sign In with Apple & Enable Apple as a sign-in provider sections found [here](https://firebase.google.com/docs/auth/ios/apple).
* I depended on the [firebase_auth_oauth](https://pub.dev/packages/firebase_auth_oauth) package
    * `import 'package:firebase_auth/firebase_auth.dart';`
    * `import 'package:firebase_auth_oauth/firebase_auth_oauth.dart';`
* I was able to retrieve a signed in FirebaseUser with the following code snippet
    *  `User user = await FirebaseAuthOAuth().openSignInFlow("apple.com", ["email"], {"locale": "en"});`

I've tested this code with Android & Web successfully. I have not yet validated with iOS, however the docs claim to support all 3 platforms.

**Reply by matteodg** [reply]
Unfortunately I tried with the iOS simulator, but after I insert the password and press continue it just hangs there.

**Reply by DennisZachotzky** [reply]
Can you try it on a real device? Sometimes the emulator causes problems

**Reply by sageata** [reply]
> Unfortunately I tried with the iOS simulator, but after I insert the password and press continue it just hangs there.

This is a very well know issue by Apple but they are 'too busy' to fix it, on top of it, your app can get rejected if they are not testing on a real device.

**Reply by michaelsof47** [reply]
i have tried it. actually, it works in iOS and Android well. but, i can't get fullName from iOS because of User.displayName (from Firebase) and fullName is different handed. Anyway, if i have a opportunity to contribute with this project. can be possible ?

**Comment by dsgriffin** [comment]
Any recent plans or updates to address this?

**Comment by enzodanjour** [comment]
+1

**Comment by rahulraj64** [comment]
@Salakar Please provide an update on this. People are desperately waiting for Apple sign in on Android.

**Comment by offline-first** [comment]
@sageata  any updates?  I have many users who want to use their Apple Sign in on their Android phone.

**Reply by sageata** [reply]
Hi @offline-first, I don't have any news, it seems that the FlutterFire team doesn't want to implement this feature or there might be some other technical limitations that they have.

Anyways, I can share with you my code so you can see what's required on both iOS and Android


**Android**

MainActivity.kt

```
package YOUR_PACKAGE_NAME

import android.util.Log
import androidx.annotation.NonNull
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.OAuthProvider
import com.google.gson.Gson
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugins.GeneratedPluginRegistrant
import org.json.JSONObject

class MainActivity : FlutterActivity() {

    private val channel = "YOUR_CHANNEL_NAME"
    private val TAG = MainActivity::class.java.simpleName

    override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
        GeneratedPluginRegistrant.registerWith(flutterEngine);

        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, channel).setMethodCallHandler { call, result ->

            if (call.method == "signInWithApple") {
                val auth = FirebaseAuth.getInstance();
                val provider = OAuthProvider.newBuilder("apple.com")
                provider.setScopes(arrayOf("email", "name").toMutableList());
                val pending = auth.pendingAuthResult;

                if (pending != null) {
                    pending.addOnSuccessListener { authResult ->
                        Log.d(TAG, "checkPending:onSuccess:$authResult")
                        authResult.user
                        val user = authResult.user;
                        result.success(user?.uid);
                    }.addOnFailureListener { e ->
                        Log.w(TAG, "checkPending:onFailure", e)
                        result.error("checkPending:onFailure", e.toString(), e.toString());
                    }
                } else {
                    Log.d(TAG, "pending: null")
                    auth.startActivityForSignInWithProvider(this, provider.build()).addOnSuccessListener { authResult ->
                        // Sign-in successful!
                        Log.d(TAG, "activitySignIn:onSuccess:${authResult.user}")
                        val user = authResult.user;
                        result.success(user?.uid);
                    }
                            .addOnFailureListener { e ->
                                Log.w(TAG, "activitySignIn:onFailure", e)
                                result.error("activitySignIn:onFailure", e.toString(), e.toString());
                            }
                }
            } else {
                result.notImplemented();
            }
        }

    }
}
```

**Dart**

```
if (Platform.isAndroid) {
          try {
            final platform = MethodChannel('YOUR_CHANNEL_NAME');
            final result = await platform.invokeMethod('signInWithApple');
            print(result);
          } on PlatformException catch (e) {
            print(e);
          }
        }
}        
```

You will have to listen to FirebaseAuth events to see if is logged in, but I think that's all the code you need to implement this on Android.

**Comment by offline-first** [comment]
Thank you @sageata this solution works great!

**Reply by sageata** [reply]
@offline-first, I'm happy I could help ;)

**Comment by VictorUvarov** [comment]
@sageata I ended up making a small android [plugin](https://github.com/Tive-technology/apple_sign_in) with your change and using the firebase auth parsing logic to return the result.

**Comment by melihhakanpektas** [comment]
```
Future<UserCredential> signInWithApple() async {
    final provider = OAuthProvider("apple.com")
      ..addScope('email')
      ..addScope('name');

    return kIsWeb
        ? await FirebaseAuth.instance.signInWithPopup(provider)
        : await FirebaseAuth.instance.signInWithProvider(provider);
  }
```
This code solved my apple login problem on android and web platforms.

**Reply by gpelizzo** [reply]
Hi, 
This is not sufficient. If you do have a backend, the id token is mandatory. On Android it does not return the id token (jwt).
